### PR TITLE
monorepo

### LINKED ISSUE #178: monorepository
# TL;DR

use [lernajs](https://lernajs.io/)

### PR DESCRIPTION
**Pull request by shoota**
# TL;DR

Closes #178

- [x] each packaged file deploy to cdn by CI service
- [x] npm publication workflow (prefer to use CI service)
- [x] dependency service migration

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -84,31 +28,19 @@
...
-    "test": "karma start --single-run"
-  },
-  "size-limit": [
-    {
-      "path": "lib/build/agent.js",
-      "limit": "8 KB"
-    }
-  ]
+    "clean": "lerna clean --yes && lerna run clean",
+    "coverage": "lerna run coverage",
```
Comment (shoota) [inline package.json]: `run <script> [args..]    Run an npm script in each package that contains that script.`

File: package.json [hunk 2]
```diff
@@ -84,31 +28,19 @@
...
-    "prepare": "npm run build",
-    "test": "karma start --single-run"
-  },
-  "size-limit": [
-    {
-      "path": "lib/build/agent.js",
-      "limit": "8 KB"
-    }
-  ]
+    "clean": "lerna clean --yes && lerna run clean",
```
Comment (shoota) [inline package.json]: `lerna clean` will remove `node_modules` in each package.
`lerna run clean` will execute `npm run clean` in each package.

**Comment by shoota** [comment]
Not fix to deploy and publication yet.

**Comment by shoota** [comment]
dpendabot will be support lerna.
https://github.com/dependabot/dependabot-core/pull/197

File: .travis.yml [hunk 3]
```diff
@@ -14,29 +14,23 @@ script:
 - npm run $CMD
 after_success:
 - '[ "$CMD" == "test" ] && npm run coverage'
+- 'if [ "${TRAVIS_TAG}" != "" ] && [ "$CMD" = "ci:sauce" ]; then
+      npm run publish;
```
Comment (9renpoto) [inline .travis.yml]: 🤔

**Comment by shoota** [comment]
### publication work flow
- run `npm run bump` at local
  - bump up version each changed package
  - make git tag commit
- push tag to remote repository
  - run CI and execute `lerna publish --skip-git`
  - publishing to npm with tagged version

File: packages/agent/robots.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+User-agent: *
```
Comment (9renpoto) [inline packages/agent/robots.txt]: move to rootDir

File: cdn/agent.d.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+/*! @userdive/agent 0.11.0-beta | Copyright (c) 2018 UNCOVER TRUTH Inc. | License GPL-3.0 */
```
Comment (9renpoto) [inline cdn/agent.d.js]: please remove build files

File: .travis.yml [hunk 6]
```diff
@@ -11,6 +11,8 @@ env:
   - CMD=ci:sauce
   - CMD=lint
   - CMD=test
+  global:
+    secure: eMzPfJSrLuDVCILlalM1QY7cuQn8PjZuKtY3u7y/3EUvb5XKg0TojJSyDGQNKuHCSonSByuR5AP68VzCZBC5YSWiTwSCRysppr5/PyQcojRTEehET+AOS7MWlZprDsrF3KsP5OmLS4fbVIkD/FklIOxCvALcSGmN/ZIavPk42dYcKAd63oYjijR/34VXyBaQ/LEhwgnsQ+arKAiuHCc0KADbw+qnqKuMNwBIv/VzNJdMLK+8qPew0AEhQSEZm8v1ihaobLqKq+Oif43ft8ldS6L92kTnZCaCFgGPUGa398E9FJcqcct5m+I2A6T9VN0AL8RB61T7r/M9ow6sYTQW5KnunX+ZPGKergtP9yrmd9PaMv5xr+vMIYUTyw2bQos8f/hkXd+BXS0V5PvpQ1aLZZH0YYx++GFAMRdVnBfZaMKXrH0VWyHlt4TRMxab7PmuWE5w/vymCl29DHLdiLO9syM4rvvrVXJVe1hw2rjuNsprp4KSbCrDlgvkIsql5BourjA8pOeP8U56Mo/PPtNLbgza4DNqmtu+pdg8Qh1Xd6i7/bBYJRBQuHjuXTLy60r42rZvnkO0TINv47HUyzed0DM3DnaWnYlvkV89ZQzaMAa7z77fPYbB5dTzFrGGOvtGg/kJd/sVyI2ojQ7elK88tU6/Wk+vX2Hlff/INOkIQJE=
```
Comment (shoota) [inline .travis.yml]: npm token

File: .travis.yml [hunk 7]
```diff
@@ -22,11 +24,9 @@ notifications:
 before_deploy:
 - npm run build
 deploy:
-- provider: npm
+- provider: script
+  script: script/publish
```
Comment (shoota) [inline .travis.yml]: run publication script

File: packages/agent/package.json [hunk 8]
```diff
@@ -0,0 +1,103 @@
+{
+  "name": "@userdive/agent",
+  "description": "webpage tracking agent",
+  "version": "0.11.0-beta",
```
Comment (9renpoto) [inline packages/agent/package.json]: old?

Comment (shoota) [reply packages/agent/package.json]: oh...

File: packages/agent/tsconfig.json [hunk 9]
```diff
@@ -21,7 +21,8 @@
     "noUnusedLocals": true,
     "noUnusedParameters": true,
     "moduleResolution": "node",
-    "outDir": "lib"
+    "outDir": "lib/js",
+    "declarationDir": "lib/typings"
```
Comment (shoota) [inline packages/agent/tsconfig.json]: transpile to `lib/js`, generate definition to `lib/typings`. 
`lib/` dir publishes to npm, `lib/build` dir put to CDN.

File: packages/agent/package.json [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "homepage": "https://github.com/userdive/agent.js#readme",
+  "keywords": [
+    "userdive"
+  ],
+  "license": "GPL-3.0",
+  "main": "lib/agent.js",
+  "types": "lib/typings/agent.d.ts",
```
Comment (shoota) [inline packages/agent/package.json]: added 'types'. (plugins will need agent declaration)

**Comment by shoota** [comment]
### publication work flow
- run `npm run bump` at local **release branch**
  - bump up version each changed package
- check bump packages, commit & push it.
- merge release branch on remote.
- make tag at local **master branch**
- push tag to **remote master**
  - run CI and execute `lerna publish --skip-git` via script/publish
  - publishing to npm with tagged version

**Pull request merged**